### DISCUSSION TITLE
Python app calls a native command line tool - how to bundle?

### DISCUSSION THREAD

**Discussion by clstaudt** [pyinstaller/pyinstaller#7100] [Help]
My Python app calls  [wkhtmltopdf](https://wkhtmltopdf.org) via [pdfkit](https://github.com/JazzCore/python-pdfkit).

Apparently this does no longer work when the bundled app is started by double click and not from the terminal.

What would be the correct way to enable this?

**Comment by rokm** [comment]
What OS is that? Can you provide a minimal example?

**Reply by clstaudt** [reply]
macOS, Linux, and Windows. I would need some time to work out a minimal example - meanwhile, this is the demo app: https://github.com/tuttle-dev/tuttle/releases

The general question is whether it is possible to not only bundle Python libraries but also native tools which are called by the Python libraries. It would not be practical to have the user install the tool separately.

**Reply by bwoodsend** [reply]
The short answer is yes. You can `--add-data` an executable and launch it via `subprocess`. You should ideally vendor it inside your `tuttle` package, locating it relative to `tuttle.__file__` then it more or less ceases to be a PyInstaller problem.

**Reply by rokm** [reply]
That depends on the tool itself - whether it is portable (has no hardcoded absolute paths) and whether it is self-contained enough to be collected into your application bundle without having to collect half the OS with it. It also depends on how the python library finds the tool.

With `pdfkit`, the following seems to work on Windows:
```python
# program.py
import sys
import os
if getattr(sys, 'frozen', False):
    os.environ['PATH'] = sys._MEIPASS + os.pathsep + os.environ['PATH']
import pdfkit
pdfkit.from_url('http://google.com', 'out.pdf')
```
```
pyinstaller ^
  --clean  ^
  --noconfirm  ^
  program.py  ^
  --add-binary c:\Users\Rok\Downloads\wkhtmltox\bin\wkhtmltoimage.exe;.  ^
  --add-binary c:\Users\Rok\Downloads\wkhtmltox\bin\wkhtmltox.dll;.  ^
  --add-binary c:\Users\Rok\Downloads\wkhtmltox\bin\wkhtmltopdf.exe;.
```

I.e., we collect `wkhtmltox` to top-level application dir, and set `PATH` accordingly in the program before trying to import `pdfkit`, so it can [discover the executable via `where.exe`](https://github.com/JazzCore/python-pdfkit/blob/c83fa250f85b210f4a0f05ca613ec0fb9580732e/pdfkit/configuration.py#L25-L26).

**Reply by rokm** [reply]
I wonder how trying to run `which` behaves within an macOS app bundle, though (notwithstanding the limited environment variables when running the bundle via double-click).
